### PR TITLE
[DeepSeek] all-to-all-v kernel writes out `output_splits`

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #956
* #989
* #958
* #954
* #952
* __->__ #941

This PR contains three changes:
    
    1. all-to-all-v kernel writes out `output_splits` instead of total received
    (so that we can use the split info as `input_splits` in the second shuffle or backward)
    
    2. Add backward for all_to_all_v (WIP)
      
    3. Move parts of the model out of `no_grad()`

ghstack-source-id: 8e0dda40b4dc9206343a020c42e8cd8c5741f860
Pull Request resolved: https://github.com/pytorch/torchtitan/pull/941
    
Test
```
$ torchrun --standalone --nproc-per-node 4 run.py
...
logits: torch.Size([2, 128, 102400])
logits: torch.Size([2, 128, 102400])
loss=tensor(510.3908, device='cuda:2', grad_fn=<MeanBackward0>)
loss=tensor(506.3376, device='cuda:3', grad_fn=<MeanBackward0>)
Backward done
Backward done
Backward done
Backward done
```

cc @yifuwang @ngimel

### CODE REVIEW

File: torchtitan/experiments/deepseek_v3/model.py [hunk 1]
```diff
@@ -562,26 +562,22 @@ def moe_infer(self, x, topk_ids, topk_weight):
...
                 group=self.ep_group,
             )
+            # Output splits sanity check
+            # expected_splits = torch.sum(
+            #     tokens_per_expert_group.view(self.ep_size, -1),
+            #     dim=1,
+            # )
+            # torch.testing.assert_close(self.output_splits, expected_splits)
             # Received tokens from all other ranks. TODO: use mask instead
+            received = self.output_splits.sum().item()
```
Comment (ngimel) [inline torchtitan/experiments/deepseek_v3/model.py]: this is an h2d sync?

Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/model.py]: It is. I am looking for an mechanism that can consume the `output_splits` directly.
Would your PR be related? https://github.com/pytorch/pytorch/pull/148531
Without that mechanism, I am putting a sync here for now.

Comment (ngimel) [reply torchtitan/experiments/deepseek_v3/model.py]: It's related but it comes with "all splits should be divisible by 16" limitation, for inference maybe this grouped gemm https://github.com/pytorch/FBGEMM/blob/98e9e680c664421a6157d55079e58c7938452798/fbgemm_gpu/experimental/gemm/triton_gemm/grouped_gemm.py#L131 is better.

**Comment by yifuwang** [comment]
Looks good. Though you probably want to return both `received` and `output_split_sizes`. This doesn't avoid the CPU sync if you need `received` on the host, but avoids a `sum` kernel launch.

**Comment by kwen2501** [comment]
@yifuwang Good idea. I hope I can find a way to **not** used the summed `received` entirely in some later PRs. So will keep things as is for now.

File: torchtitan/experiments/deepseek_v3/model.py [hunk 2]
```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
...
         hidden_states = hidden_states.view(-1, hidden_states.shape[-1])
-        y = self.moe_infer(hidden_states, topk_idx, topk_weight).view(*orig_shape)
+        y = self.moe_forward(hidden_states, topk_idx, topk_weight).view(*orig_shape)
         if self.config.n_shared_experts is not None:
             y = y + self.shared_experts(identity)
         return y
 
-    @torch.no_grad()
-    def moe_infer(self, x, topk_ids, topk_weight):
+    def moe_forward(self, x, topk_ids, topk_weight):
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:533]: would be nice to add docstring to moe_forward (well, probably to other methods too) that document the shapes of the input tensors, e.g. `topk_ids: [seq_len, k]`

File: torchtitan/experiments/deepseek_v3/model.py [hunk 3]
```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
...
-        tokens_per_expert = cnts.sum(dim=0)
-        # Token indices for each expert
-        idxs = topk_ids.view(-1).argsort()
-        sorted_tokens_shape = idxs.shape + x.shape[1:]
+        # This part sorts the token indices so that tokens routed to the same expert reside consecutively.
+        # An implication is that tokens to the same "expert group" (i.e., device) are also consecutive.
+        # Since this is an "aritificial" index creation (final outcome being
+        # `idxs`), we don't need gradients here.
+        with torch.no_grad():
+            # [seq_len, n_routed_experts]
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:543]: maybe collaps the 3 lines (cnts=, cnts.scatter_, tokens_per_expert) into a group of 3 and give one comment above them that says 
`# form a one_hot encoding of shape (seq_len, experts) and then sum to get #tokens per expert`
and maybe rename `tokens_per_expert` to `num_tokens_per_expert` or something?

File: torchtitan/experiments/deepseek_v3/model.py [hunk 4]
```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
...
+            cnts = topk_ids.new_zeros((topk_ids.shape[0], len(self.experts)))
+            # Fill 1 to the selected experts
+            cnts.scatter_(1, topk_ids, 1)
+            tokens_per_expert = cnts.sum(dim=0)
+            # Token indices for each expert
+            idxs = topk_ids.view(-1).argsort()
+            sorted_tokens_shape = idxs.shape + x.shape[1:]
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:550]: so we flatten the idxs (each sequence is appended to one long row), then we shuffle them so that tokens from different sequences that are going to the same expert are next to each other in memory.

and x.shape is what, (b*s,h)? if so, then i think it makes sense that sorted_tokens_shape would be (b*s, h)

File: torchtitan/experiments/deepseek_v3/model.py [hunk 5]
```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
...
+
         # Take necessary space from the `token_send_buf` symm mem
         sorted_tokens = self.token_send_buf[: idxs.shape[0]]
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:553]: nit: is `[: idxs.shape[0]]` broken formatting? i thought it would be :idxs... (no space) - otherwise looked like it was 2 dim and missing comma.

File: torchtitan/experiments/deepseek_v3/model.py [hunk 6]
```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
...
+        sorted_tokens.copy_(x[idxs // topk_ids.shape[1]])
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:554]: x: [b*s, h]?
idxs: [b*s]?
topk_ids: [s, k]

idxs represents "the indices of tokens respective to b*s flattening, such that the tokens for expert0 come first, etc", right?

i'm confused about what `idxs // topk_ids.shape[1]` is doing.

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py [hunk 7]
```diff
@@ -156,3 +158,34 @@ def on_device_all_to_all_v(
...
+
+class OnDeviceAllToAllV(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        output: torch.Tensor,
+        output_splits: torch.Tensor,
+        input: torch.Tensor,
+        input_splits: torch.Tensor,
+        group: dist.ProcessGroup = dist.group.WORLD,
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py:171]: should we default this? it feels like usually you need to know what group you're doing this over. i could see leaving it as 'world' or unit testing purposes but for any real usage in models it seems unlikely the default would be sufficient?

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py [hunk 8]
```diff
@@ -156,3 +158,34 @@ def on_device_all_to_all_v(
...
+        ctx.input_shape = input.shape
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        grad_output_splits = ctx.saved_tensors
+        grad_input_splits = torch.empty_like(grad_output_splits)
+        grad_input = grad_output.new_empty(*ctx.input_shape)
+        _on_device_all_to_all_v(
+            grad_input, grad_input_splits, grad_output, grad_output_splits, group=group
```
Comment (EugenHotaj) [inline torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: I might be missing something but I don't think this `backward()` function is ever being called -- otherwise `group` would fail here since it should be `ctx.group`

File: torchtitan/experiments/deepseek_v3/model.py (hunk 6 above)
Comment (wconstab) [reply torchtitan/experiments/deepseek_v3/model.py:554]: ok- the explanation is
1) idxs contains the indexes of tokens in a sequence that is K times longer than the original input sequence, due to the view(-1) above before the argsort
2) the index values logically refer to tokens in the original input sequence (x) so to make them useful again we floor-didide by `k` (the value used for topk).

File: torchtitan/experiments/deepseek_v3/model.py [hunk 9]
```diff
@@ -562,39 +573,44 @@ def moe_infer(self, x, topk_ids, topk_weight):
...
+        # device). This part doesn't need gradient.
+        with torch.no_grad():
             tokens_per_expert_post_gather = tokens_per_expert_group.view(
                 self.ep_size, self.experts_per_rank
             ).sum(dim=0)
             gatherd_idxs = np.zeros(shape=(gathered_tokens.shape[0],), dtype=np.int32)
             s = 0
-            for i, k in enumerate(tokens_per_expert_group.cpu().numpy()):
+            # TODO: remove `tolist()`
+            for i, k in enumerate(tokens_per_expert_group.tolist()):
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:606]: this is still a cuda-cpu sync point, right? can we avoid it by doing a scan on device or something?

File: torchtitan/experiments/deepseek_v3/model.py [hunk 10]
```diff
@@ -612,25 +628,24 @@ def moe_infer(self, x, topk_ids, topk_weight):
...
+            self.ep_group,
+        )
 
-        new_x = torch.empty_like(outs)
-        new_x[idxs] = outs
+        returned_tokens = torch.empty_like(gathered_tokens)
+        returned_tokens[idxs] = gathered_tokens
         final_out = (
-            new_x.view(*topk_ids.shape, -1)
+            returned_tokens.view(*topk_ids.shape, -1)
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model.py:648]: i'd also like to see `topk_weight` documented at the top of the function.  I am assuming that this helps to implement a weighted sum for combining the expert outputs into one output, but its a little hard to follow the code.

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py (hunk 8 above)
Comment (wconstab) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: yea good point. but it does look like the on_device all2allv is called in forward part of moe layer in a part that has gradients enabled.  (it is also called under 'with no_grad()` but not exclusively)

Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: @EugenHotaj @wconstab Thanks. You are right. 
I am backing out the backward function for `on_device_all_to_all_v` from this PR.
The reason is that autograd does not allow touching the buffer a second time between fwd and bwd. This conflicts with sharing the symm mem buffer across layers (to save memory) as in PR #958.

Instead, I am adding backward support via the traditional `all_to_all_single_autograd` (Yes, there is a CPU sync.) But the backward is more stable. PR is at #989.

`on_device_all_to_all_v` is now limited to inference / generation.

File: torchtitan/experiments/deepseek_v3/model.py (hunk 5 above)
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/model.py:553]: I remember the space is required by formatter too.

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py (hunk 7 above)
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py:171]: The default here is similar to torch.dist defaulting group to `None`

**Comment by kwen2501** [comment]
@wconstab thanks for all the requests for comments. I will put a PR on top of the current stack -- too hard to modify this one without resolving conflicts with other PRs on top.

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py (hunk 8 above)
Comment (wconstab) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: sounds good for short term.  longer term we'd want to fix the backward+symm memm issue though right?  can we mark the tensor as a 'buffer' to make autograd not care about it, or something?

**Review by wconstab — APPROVED** [review]

File: torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py (hunk 8 above)
Comment (ngimel) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: Autograd does not allow touching the buffer that's going to be used by autograd, is that the case here? If you still need all this memory live, then registering multiple copies with symmetric memory shouldn't be overall memory hit.

Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/symm_mem_recipes/triton_on_device_all_to_all_v.py]: @wconstab @ngimel yes. 
@EugenHotaj and @yifuwang suggested something similar -- we can try moving the symm mem inside an autograd function, so that it is not visible to the autograd engine.
